File: Linked List/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
       dummy = ListNode(-1)
       dummy.next = head
       slow, fast = dummy, dummy
       for _ in range(n):
           fast = fast.next
       while fast.next:
           slow = slow.next
           fast = fast.next
       slow.next = slow.next.next
       return dummy.next

# @lc code=end

File: Linked List/19.remove-nth-node-from-end-of-list.cpp
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
       auto dummy = new ListNode(-1);
       dummy->next = head;
       auto first = dummy, second =dummy;
       while(n--) first = first->next;
       while(first->next)
       {
           first = first->next;
           second = second->next;
       }
       second->next = second->next->next;
       return dummy->next;
    }
};
// @lc code=end

